lang/java/compiler/src/main/java/org/apache/avro/compiler/idl/DocCommentHelper.java

```diff
@@ -0,0 +1,155 @@
+ */
+package org.apache.avro.compiler.idl;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Utility class with {@code ThreadLocal} fields that allow the generated
+ * classes {@link Idl} and {@link IdlTokenManager} to exchange documentation
+ * comments without forcing explicit parsing of documentation comments.
+ *
+ * The reason this works is that all calls to this class happen within a call to
+ * the method {@link Idl#CompilationUnit()} (either directly or indirectly).
+ */
+public class DocCommentHelper {
+  /**
+   * Pattern to match the common whitespace indents in a multi-line String.
+   * Doesn't match a single-line String, fully matches any multi-line String.
+   *
+   * To use: match on a {@link String#trim() trimmed} String, and then replace all
+   * newlines followed by the group "indent" with a newline.
+   */
+  private static final Pattern WS_INDENT = Pattern.compile("(?U).*\\R(?<indent>\\h*).*(?:\\R\\k<indent>.*)*");
+  /**
+   * Pattern to match the whitespace indents plus common stars (1 or 2) in a
+   * multi-line String. If a String fully matches, replace all occurrences of a
+   * newline followed by whitespace and then the group "stars" with a newline.
+   *
+   * Note: partial matches are invalid.
+   */
+  private static final Pattern STAR_INDENT = Pattern.compile("(?U)(?<stars>\\*{1,2}).*(?:\\R\\h*\\k<stars>.*)*");
+
+  private static final ThreadLocal<DocComment> DOC = new ThreadLocal<>();
+  private static final ThreadLocal<List<String>> WARNINGS = ThreadLocal.withInitial(ArrayList::new);
+
+  /**
+   * Return all warnings that were encountered while parsing, once. Subsequent
+   * calls before parsing again will return an empty list.
+   */
+  static List<String> getAndClearWarnings() {
+    List<String> warnings = WARNINGS.get();
+    WARNINGS.remove();
+    return warnings;
+  }
+
+  static void setDoc(Token token) {
+    DocComment newDocComment = new DocComment(token);
+    DocComment oldDocComment = DOC.get();
+    if (oldDocComment != null) {
+      WARNINGS.get().add(String.format(
+          "Found documentation comment at line %d, column %d. Ignoring previous one at line %d, column %d: \"%s\"\n"
+              + "A common cause is to use documentation comments ( /** ... */ ) instead of multiline comments ( /* ... */ ).",
+          newDocComment.line, newDocComment.column, oldDocComment.line, oldDocComment.column, oldDocComment.text));
+    }
+    DOC.set(newDocComment);
+  }
+
+  static void clearDoc() {
+    DocComment oldDocComment = DOC.get();
+    if (oldDocComment != null) {
+      WARNINGS.get().add(String.format("Ignoring out-of-place documentation comment at line %d, column %d: \"%s\"\n"
+          + "A common cause is to use documentation comments ( /** ... */ ) instead of multiline comments ( /* ... */ ).",
+          oldDocComment.line, oldDocComment.column, oldDocComment.text));
+    }
+    DOC.remove();
+  }
```

Comment:
This method is not used yet in `idl.jj`, but is in preparation of ignoring misplaced documentation (for example halfway a `fixed` type declaration).